pinot-controller/src/main/java/org/apache/pinot/controller/workload/scheme/PropagationUtils.java

```diff
@@ -101,74 +147,129 @@ private static void collectHelixTagsForTable(List<String> tags, TenantConfig ten
+   * @param pinotResourceManager Resource manager to fetch table configs.
+   * @param tableName The raw or type-qualified table name.
+   * @return A list of Helix tags associated with the table.
    */
   public static List<String> getHelixTagsForTable(PinotHelixResourceManager pinotResourceManager, String tableName) {
+    if (tableName == null || tableName.trim().isEmpty()) {
+      throw new IllegalArgumentException("Table name cannot be null or empty");
+    }
+
     List<String> combinedTags = new ArrayList<>();
     TableType tableType = TableNameBuilder.getTableTypeFromTableName(tableName);
     List<String> tablesWithType = (tableType == null)
         ? Arrays.asList(TableNameBuilder.OFFLINE.tableNameWithType(tableName),
             TableNameBuilder.REALTIME.tableNameWithType(tableName))
         : Collections.singletonList(tableName);
     for (String table : tablesWithType) {
-      TableConfig tableConfig = pinotResourceManager.getTableConfig(table);
-      if (tableConfig != null) {
-        collectHelixTagsForTable(combinedTags, tableConfig.getTenantConfig(), tableConfig.getTableType());
+      try {
+        TableConfig tableConfig = pinotResourceManager.getTableConfig(table);
+        if (tableConfig != null && tableConfig.getTenantConfig() != null && tableConfig.getTableType() != null) {
+          collectHelixTagsForTable(combinedTags, tableConfig.getTenantConfig(), tableConfig.getTableType());
+        }
+      } catch (Exception e) {
+        throw new RuntimeException("Failed to get table config for table: " + table, e);
       }
     }
     return combinedTags;
   }
 
   /**
-   * Get the mapping between helix tag -> instances
+   * Builds a mapping from Helix tag to the set of instances carrying that tag.
+   *
+   * @param pinotResourceManager Resource manager used to fetch instance configs.
+   * @return A mapping of Helix tag → set of instance names.
    */
   public static Map<String, Set<String>> getHelixTagToInstances(PinotHelixResourceManager pinotResourceManager) {
     Map<String, Set<String>> tagToInstances = new HashMap<>();
-    for (InstanceConfig instanceConfig : pinotResourceManager.getAllHelixInstanceConfigs()) {
-      String instanceName = instanceConfig.getInstanceName();
-      for (String helixTag : instanceConfig.getTags()) {
-        tagToInstances.computeIfAbsent(helixTag, tag -> new HashSet<>()).add(instanceName);
+    try {
+      List<InstanceConfig> instanceConfigs = pinotResourceManager.getAllHelixInstanceConfigs();
+      if (instanceConfigs == null) {
+        LOGGER.warn("No instance configs found, returning empty mapping");
+        return tagToInstances;
       }
+
+      for (InstanceConfig instanceConfig : instanceConfigs) {
+        if (instanceConfig == null) {
+          LOGGER.warn("Skipping null instance config");
+          continue;
+        }
+        String instanceName = instanceConfig.getInstanceName();
+        if (instanceName == null || instanceName.trim().isEmpty()) {
```

Comment:
I see a number of additional validation added. Are these all necessary? For example, can there be an instance with null name or empty name?